tsdb/shard.go

```diff
@@ -1959,48 +2096,197 @@ func (fs *MeasurementFieldSet) marshalMeasurementFieldSetNoLock() (marshalled []
+		}
+	}
+	return b, nil
+}
+
+func (fs *MeasurementFieldSet) load() (rErr error) {
+	err := func() error {
+		fs.mu.Lock()
+		defer fs.mu.Unlock()
+
+		pb, err := fs.loadParseFieldIndexPB()
+		if err != nil {
+			return err
+		}
+		fs.fields = make(map[string]*MeasurementFields, len(pb.GetMeasurements()))
+		for _, measurement := range pb.GetMeasurements() {
+			fields := make(map[string]*Field, len(measurement.GetFields()))
+			for _, field := range measurement.GetFields() {
+				fields[string(field.GetName())] = &Field{Name: string(field.GetName()), Type: influxql.DataType(field.GetType())}
+			}
+			set := &MeasurementFields{}
+			set.fields.Store(fields)
+			fs.fields[string(measurement.GetName())] = set
+		}
+		return nil
+	}()
+
+	if err != nil {
+		return fmt.Errorf("failed loading field indices: %w", err)
+	}
+	return fs.ApplyChanges()
+}
+
+func (fs *MeasurementFieldSet) loadParseFieldIndexPB() (pb *internal.MeasurementFieldSet, rErr error) {
+	pb = &internal.MeasurementFieldSet{}
 
 	fd, err := os.Open(fs.path)
 	if os.IsNotExist(err) {
-		return nil
+		return pb, nil
 	} else if err != nil {
-		return err
+		err = fmt.Errorf("failed opening %s: %w", fs.path, err)
+		return nil, err
 	}
-	defer fd.Close()
+
+	defer errors2.Capture(&rErr, func() error {
+		if e := fd.Close(); e != nil {
+			return fmt.Errorf("failed closing %s: %w", fd.Name(), e)
+		} else {
+			return nil
+		}
+	})()
 
 	var magic [4]byte
 	if _, err := fd.Read(magic[:]); err != nil {
-		return err
+		err = fmt.Errorf("failed reading %s: %w", fs.path, err)
+		return nil, err
 	}
 
 	if !bytes.Equal(magic[:], fieldsIndexMagicNumber) {
-		return ErrUnknownFieldsFormat
+		return nil, fmt.Errorf("%q: %w", fs.path, ErrUnknownFieldsFormat)
 	}
 
-	var pb internal.MeasurementFieldSet
 	b, err := io.ReadAll(fd)
 	if err != nil {
-		return err
+		err = fmt.Errorf("failed reading %s: %w", fs.path, err)
+		return nil, err
+	}
+	if err = proto.Unmarshal(b, pb); err != nil {
+		err = fmt.Errorf("failed unmarshaling %s: %w", fs.path, err)
+		return nil, err
 	}
+	return pb, err
+}
+
+func (fscm *measurementFieldSetChangeMgr) loadAllFieldChanges(log *zap.Logger) (changes []FieldChanges, rErr error) {
+	var fcs FieldChanges
 
+	fscm.mu.Lock()
+	defer fscm.mu.Unlock()
+	fd, err := os.Open(fscm.Path)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		err = fmt.Errorf("failed opening %s: %w", fscm.Path, err)
+		log.Error("field index file of changes", zap.Error(err))
+		return nil, err
+	}
+	defer errors2.Capture(&rErr, func() error {
+		if e := fd.Close(); e != nil {
+			return fmt.Errorf("failed closing %s: %w", fd.Name(), e)
+		} else {
+			return nil
+		}
+	})()
+	for fcs, err = fscm.loadFieldChangeSet(fd); err == nil; fcs, err = fscm.loadFieldChangeSet(fd) {
+		changes = append(changes, fcs)
+	}
+	if errors.Is(err, io.EOF) {
+		return changes, nil
+	} else if errors.Is(err, io.ErrUnexpectedEOF) {
```

Comment:
Would be good to have a test for this read behaviour.

Reply:
Test implemented.

Reply:
`TestMeasurementFieldSet_CorruptChangeFile`